apis/projectcontour/v1/helpers.go

```diff
@@ -78,3 +80,193 @@ func (r *Route) AuthorizationContext(parent map[string]string) map[string]string
+func (dc *DetailedCondition) AddWarning(warnType, reason, message string) {
+
+	detailedReason := warnType + reason
+	dc.updateReason(detailedReason, message)
+
+	i := getIndex(warnType, dc.Warnings)
+
+	if i == -1 {
+		dc.Warnings = append(dc.Warnings, SubCondition{
+			Type:    warnType,
+			Status:  ConditionTrue,
+			Reason:  reason,
+			Message: message,
+		})
+		return
+	}
+
+	// If the warning is already present, just update the reason and message.
+	dc.Warnings[i].updateReason(reason, message)
+}
+
+// AddWarningf adds an warning-level Subcondition to the DetailedCondition, using
+// fmt.Sprintf on the formatmsg and args params.
+// If a SubCondition with the given errorType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddWarningf(warnType, reason, formatmsg string, args ...interface{}) {
+
+	dc.AddWarning(warnType, reason, fmt.Sprintf(formatmsg, args...))
+}
+
+// GetWarning gets an warning of the given warnType.
+// Similar to a hash lookup, will return true in the second value if a match is
+// found, and false otherwise.
+func (dc *DetailedCondition) GetWarning(warnType string) (SubCondition, bool) {
+
+	i := getIndex(warnType, dc.Warnings)
+
+	if i == -1 {
+		return SubCondition{}, false
+	}
+
+	return dc.Warnings[i], true
+}
+
+// updateReason updates a DetailedCondition's reason and message correctly
+// if they are different to the existing ones.
+func (dc *DetailedCondition) updateReason(reason, message string) {
+	if dc.Reason == "" {
+		dc.Reason = reason
+		dc.Message = message
+		return
+	}
+
+	if dc.Reason != reason {
+		dc.Reason = "MultipleProblems"
+		dc.Message = "Multiple problems were found, see errors for details"
+		return
+	}
+
+	if dc.Message != message {
+		dc.Message = dc.Message + ", " + message
+	}
+}
+
+func (sc *SubCondition) updateReason(reason, message string) {
+
+	if sc.Reason == "" {
+		sc.Reason = reason
+		sc.Message = message
+		return
+	}
+
+	if sc.Reason != reason {
+		sc.Reason = "MultipleReasons"
+	}
+
+	sc.Message = sc.Message + ", " + message
+
+}
+
+// getIndex checks if a SubCondition of type condType exists in the
+// slice, and returns its index if so. If not, returns -1.
+func getIndex(condType string, subconds []SubCondition) int {
+
+	for i, cond := range subconds {
+		if cond.Type == condType {
+			return i
+		}
+	}
+	return -1
+}
+
+// GetConditionIndex gets the index of a condition of the given type from
+// the given DetailedCondition slice, or -1 if it is not found.
+func GetConditionIndex(condType string, conds []DetailedCondition) int {
+
+	for i, cond := range conds {
+		if cond.Type == condType {
+			return i
+		}
+	}
+
+	return -1
+
+}
+
+// IsPositivePolarity returns true if the DetailedCondition is a positive-polarity
+// condition like `Valid` or `Ready`, and false otherwise.
+func (dc *DetailedCondition) IsPositivePolarity() bool {
+
+	switch dc.Type {
```

Comment:
No, this is indicating if the Condition is a positive or negative polarity. In this case, "Valid" is a positive-polarity condition, meaning that if it has `Status: true`, then there is no problem, and `Status: false: means there is a problem. Other Conditions are expected to have the negative polarity, `Status: true` means there is a problem, `Status: false` means that there is no problem. Negative polarity conditions are effectively error conditions, like `InsufficientDiskSpace` or something.

Reply:
The method returns `true` if you pass it `Valid` otherwise it returns false. How then do you have a condition of `Valid` with status of `false` as described in the API comment noted above? You'll also have status `true`. 

Is the comment incorrect?

Reply:
Where this method is used is in the `AddError` method. It's used to determine what to set the condition status to - for 'Valid', it's true, which means that the status is set to `false` when an error is added. If the condition is not positive polarity, then the status is set to `true` when an error is added.

This is not setting the status directly, but is being used when an error is added to determine what the status should be set to.